core/src/smartcontracts/isi/permissions/mod.rs

```diff
@@ -2,105 +2,232 @@
+    /// [`Instruction`] operation
+    Instruction(Instruction),
+    /// [`QueryBox`] operation
+    Query(QueryBox),
+    /// [`Expression`] operation
+    Expression(Expression),
+}
+
+impl NeedsPermission for NeedsPermissionBox {
+    fn required_validator_type(&self) -> ValidatorType {
+        match self {
+            NeedsPermissionBox::Instruction(_) => ValidatorType::Instruction,
+            NeedsPermissionBox::Query(_) => ValidatorType::Query,
+            NeedsPermissionBox::Expression(_) => ValidatorType::Expression,
+        }
+    }
+}
+
+/// Implementation of this trait provides custom permission checks for the Iroha-base
+pub trait IsAllowed: Debug {
+    /// Type of operation to be checked
+    type Operation: NeedsPermission;
 
-/// Type of object validator can check
+    /// Check if the `authority` is allowed to perform `instruction`
+    /// given the current state of `wsv`.
+    ///
+    /// # Denial reasons
+    /// If the execution of `instruction` under given `authority` with
+    /// the current state of `wsv` is disallowed.
+    fn check(
+        &self,
+        authority: &AccountId,
+        operation: &Self::Operation,
+        wsv: &WorldStateView,
+    ) -> ValidatorVerdict;
+}
+
+/// Box with dyn type implementing [`IsAllowed`]
+pub type IsOperationAllowedBoxed<O> = Box<dyn IsAllowed<Operation = O> + Send + Sync>;
+
+/// Type of validator
 #[derive(Debug, Copy, Clone, PartialEq, Eq, derive_more::Display, Encode, Decode, IntoSchema)]
 pub enum ValidatorType {
-    /// [`Instruction`] variant
+    /// Validator checking [`Instruction`]
     Instruction,
-    /// [`QueryBox`] variant
+    /// Validator checking [`QueryBox`]
     Query,
-    /// [`Expression`] variant
+    /// Validator checking [`Expression`]
     Expression,
 }
-pub mod error {
-    //! Contains errors structures
 
-    use std::{convert::Infallible, str::FromStr};
+/// Verdict returned by validators
+#[derive(Debug, Clone, PartialEq, Eq, derive_more::Display, Encode, Decode, IntoSchema)]
+pub enum ValidatorVerdict {
+    /// Deny the execution of an operation and provide the [`DenialReason`].
+    ///
+    /// Something went wrong and the validator voted to deny the execution of the instruction.
+    Deny(DenialReason),
+    /// Skip an operation.
+    ///
+    /// The validator votes to skip an operation if it is not supported by the validator
+    /// or has no meaning in a particular context.
+    Skip,
+    /// Allow the execution of an instruction.
+    ///
+    /// The validator allows an instruction to be executed if
+    /// the operation is correct from its point of view.
+    Allow,
+}
+
+impl PartialOrd for ValidatorVerdict {
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(Ord::cmp(self, other))
+    }
+}
+
+// Deny < Skip < Allow
+impl Ord for ValidatorVerdict {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.as_u8().cmp(&other.as_u8())
```

Comment:
I'd rewrite this using the above comment. You won't need the comment and will be guaranteed that the comparison works regardless of architecture.

Reply:
Alright. The current implementation looks good.